### DISCUSSION TITLE
How/where will we store our data?

### DISCUSSION THREAD

**Discussion by jsadler2** [USGS-R/drb-do-ml#2] [Questions]
I'm wondering what the best place is to store our data.

I think the options are 
1. locally
2. Caldera
3. CHS S3

**Comment by jsadler2** [comment]
My vote currently is currently Caldera. I don't like "local" b/c things can get out of sync. I'm open to S3, but I'm not sure that there are many advantages over Caldera. I'd like to hear your thoughts.

**Reply by aappling-usgs** [reply]
I agree with not-local and agree that Caldera and S3 are roughly equivalent. The PUMP data-driven salinity projects are both turning to S3, I think, which could be relevant if a single tool would be easier for Lauren and Galen...but then again, there'll be some HPC activity in those projects as well.

**Reply by jds485** [reply]
I think after discussion with Sam yesterday, Caldera might be the preference over S3. Caldera is free, can use Globus and rsync to interface with locally stored files, and is ready for use with HPCs when it's time to train models. One of the lingering questions is how we handle version control (which datasets are used for which models). Here's a [link](https://doimspp.sharepoint.com/:w:/r/sites/usgs-GS-WMA-IWP-PUMP/_layouts/15/Doc.aspx?sourcedoc=%7B31640787-95FE-4AA0-BA08-DB05D879C3FA%7D&file=211021_DataStorageDiscussion.docx&action=default&mobileredirect=true) to that discussion doc.

**Reply by galengorski** [reply]
It seems like this conversation is happening in quite a few places in parallel. For the estuary salinity project I think we are going to use S3 for storage. Below is a chat from Amelia laying out some of the advantages:

>   I personally think the simpler the data system is, and the fewer copies of data you have, the better. If you have a bunch of copies of the data (Caldera, Sharepoint, local copy for developement), it is possible that they get out of sync if you are not diligent about syncing things. If you end up have 3 different versions of data in the three locations, you can get into a hairy problem to untangle.

>  I am advocating for S3 because I think we should be able to work on it as our single source of truth that we all directly read from and write to in our code. Then we don't have to worry about syncing and making sure everyone has the latest copy of the data because there is only one copy of the data.

> That being said, you can certainly make a Caldera/Sharepoint/local system work. You just need to think through the workflow for keeping things up to date. You should know what the 'master' copy of the data is, that everyone keeps up to date. And every other data storage location should have a specific purpose and defined workflow for syncing to the master.

> I still need to figure out our exact workflow with S3, but I think we should be able to read and write directly to S3 from our local machines and HPC. I need to test it out to be completely sure, but that's on my to-do list for this week.

**Reply by jsadler2** [reply]
@galengorski and @lekoenig - are you both comfortable working on Tallgrass and storing our data in Caldera?

**Reply by jsadler2** [reply]
I think a possible barrier to Tallgrass/Caldera is that you can't use your local development environment (e.g., Rstudio). There is a way to use RStudio from the browser, however, which may be workable.

The advantages I see to Tallgrass/Caldera are 1) the data is in a shared, central location, so there's no chance of it getting out of date and 2) we would be doing our computation next to the data (on Tallgrass). If we used S3, we would be having to upload and download to the S3 bucket because the computation would be done either locally or on HPC which is not on the same machine as the data.

**Reply by jzwart** [reply]
See the bottom of this [README ](https://code.usgs.gov/wma/wp/pump-temperature)for some instructions for how to launch Rstudio Server or Jupyter Lab from HPC resources. @wdwatkins  [blog ](https://code.usgs.gov/wwatkins/hpc_container_blog)has more details for how to do this. 

I'm also interested in trying out Globus SDK (https://globus-sdk-python.readthedocs.io/en/stable/) for scripting files transfers across multiple endpoints. Because USGS has a Globus subscription this would be doable. Jeff Falgout hasn't tried using Globus SDK but is happy to help get a use case set up. There's a lot of material in Globus SDK so might need a dedicated person to do a deep dive to see if we could make use of it - from what I've seen, it seems promising

**Reply by jds485** [reply]
Just curious what the value of Globus SDK vs. the browser Globus service would be. Pipeline integration and documenting what was moved come to mind, but maybe there are other benefits.

**Reply by lekoenig** [reply]
> @galengorski and @lekoenig - are you both comfortable working on Tallgrass and storing our data in Caldera?

@jsadler2 Yes, this plan sounds fine to me. I haven't used Tallgrass/Caldera yet, so I don't have firm preferences on data storage options for HPC. I'm also open to this plan even if the salinity project uses S3 since gaining experience with these different HPC configurations would help me become more familiar with the pros/cons of each.

**Reply by galengorski** [reply]
> @galengorski and @lekoenig - are you both comfortable working on Tallgrass and storing our data in Caldera?

@jsadler2 This sounds good to me, I just requested an account on tallgrass and denali, so that should be coming soon.

**Reply by jzwart** [reply]
@jds485  pipeline integration is the main benefit I was thinking of. If using dynamic branching targets (e.g. model runs, sites, etc..), it'd be quite simple to push / pull the correct files rather than doing that manually. When putting files on HPC manually in the past, I know I've occasionally forgotten which files are needed

**Reply by jsadler2** [reply]
Okay @galengorski  and @lekoenig - let's DO it! ... as in let's go with Caldera/Tallgrass. We can revisit if for some reason it's not working for us.

**Reply by galengorski** [reply]
sorry to jump in late with this, but yes tallgrass/caldera sound great. I requested an account over a week ago now, so hopefully it should be coming through pretty soon here.

**Comment by jsadler2** [comment]
Also, I'm wondering which format to store the data in. I'm partial to netcdf, personally.

**Comment by jsadler2** [comment]
### Coding independently but sharing data products

@galengorski and @lekoenig - I've been thinking about how to 
* avoid stepping on each others toes in coding, while also
* staying in sync with each others data products
 
Maybe there is an established way of doing this already. If there is, maybe @aappling-usgs you know about it??

Here's my proposed solution:
1. **for code** we all fork this repo and work on it in our own "homes" 🏡 (`/home/jsadler/`  for me) and make PRs when ready to this repo
2. **for data** we set up the output path for our pipelines (`targets` and maybe `snakemake`) to our pump folder* (`/caldera/projects/usgs/water/impd/pump/drb-do/drb-do-ml/`)
 
 
\* The data folder is _not_ a git repo and it does not contain any code. It's the  _products_ of our pipeline (the code that will be committed to this repo). It will have the same folder structure, as the git repo but will only have the `out` folders. So we will commit all of the `drb-do-ml/*/src/*` files to the repo, but all of the outputs (`drb-do-ml/*/out/*`) will go to the `/caldera/projects/usgs/water/impd/pump/drb-do/drb-do-ml/` folder. 

I think this way we can all work on code independently but work from the same data files. One other thing that we could do is add a `jsadler` or `scratch` data folder (eg, `/caldera/projects/usgs/water/impd/pump/drb-do/drb-do-ml/jsadler`) so that we can have a bit of independence on the data side too if we are doing some experimental/prototyping work and not ready for others to use the data.

note - this approach assumes that we would never want to commit any of our outputs because they will be in a different folder. If we did end up wanting to commit outputs to the repo we'd have to produce them also in a `out` (or similar) folder in the _repo_ as well as in the shared, caldera data folder.
